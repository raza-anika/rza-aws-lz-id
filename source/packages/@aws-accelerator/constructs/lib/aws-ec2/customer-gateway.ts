/**
 *  Copyright 2022 Amazon.com, Inc. or its affiliates. All Rights Reserved.
 *
 *  Licensed under the Apache License, Version 2.0 (the "License"). You may not use this file except in compliance
 *  with the License. A copy of the License is located at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 *  or in the 'license' file accompanying this file. This file is distributed on an 'AS IS' BASIS, WITHOUT WARRANTIES
 *  OR CONDITIONS OF ANY KIND, express or implied. See the License for the specific language governing permissions
 *  and limitations under the License.
 */

import * as cdk from 'aws-cdk-lib';
import { Construct } from 'constructs';

export interface CfnCustomerGatewayProps {
  /**
   * Name of the Customer Gateway
   */
  readonly name: string;

  /**
   * Gateway IP address for customer gateway
   */
  readonly ipAddress: string;

  /**
   * Gateway ASN for customer gateway
   */
  readonly bgpAsn: number;

  readonly tags?: cdk.CfnTag[];
}

interface ICustomerGateway extends cdk.IResource {
  /**
   * The identifier of the customer gateway
   *
   * @attribute
   */
  readonly customerGatewayId: string;

  /**
   * The identifier of the customer gateway
   *
   * @attribute
   */
  readonly customerGatewayName: string;
}

/**
 * Class for Customer Gateway
 */
export class CustomerGateway extends cdk.Resource implements ICustomerGateway {
  public readonly customerGatewayId: string;
  public readonly customerGatewayName: string;
  public readonly name: string;

  constructor(scope: Construct, id: string, props: CfnCustomerGatewayProps) {
    super(scope, id);
    this.name = props.name;

    const resource = new cdk.aws_ec2.CfnCustomerGateway(this, 'CustomerGateway', {
      bgpAsn: props.bgpAsn,
      ipAddress: props.ipAddress,
      type: 'ipsec.1',
      tags: props.tags,
    });
    cdk.Tags.of(this).add('Name', this.name);

    this.customerGatewayId = resource.ref;

    this.customerGatewayName = props.name;
  }
}
