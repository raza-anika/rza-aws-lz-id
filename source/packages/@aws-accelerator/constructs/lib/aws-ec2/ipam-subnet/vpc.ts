/**
 *  Copyright 2022 Amazon.com, Inc. or its affiliates. All Rights Reserved.
 *
 *  Licensed under the Apache License, Version 2.0 (the "License"). You may not use this file except in compliance
 *  with the License. A copy of the License is located at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 *  or in the 'license' file accompanying this file. This file is distributed on an 'AS IS' BASIS, WITHOUT WARRANTIES
 *  OR CONDITIONS OF ANY KIND, express or implied. See the License for the specific language governing permissions
 *  and limitations under the License.
 */
import * as AWS from 'aws-sdk';
import { IPv4CidrRange } from 'ip-num';

import { throttlingBackOff } from '@aws-accelerator/utils';

interface IVpc {
  /**
   * The CIDRs of the VPC allocated by IPAM
   */
  readonly allocatedCidrs: IPv4CidrRange[];
  /**
   * The friendly name of the VPC
   */
  readonly name: string;
  /**
   * The existing subnets in the VPC
   */
  readonly subnets: Subnet[];
  /**
   * The ID of the VPC
   */
  readonly vpcId: string;
}

interface VpcProps {
  /**
   * The base CIDR pool of the VPC
   */
  readonly basePool: string[];
  /**
   * The ID of the VPC
   */
  readonly vpcId: string;
}

export class Vpc implements IVpc {
  public allocatedCidrs: IPv4CidrRange[] = [];
  public name = '';
  public subnets: Subnet[] = [];
  public readonly vpcId: string;
  private readonly baseRanges: IPv4CidrRange[];
  private readonly ec2 = new AWS.EC2();

  constructor(props: VpcProps) {
    this.vpcId = props.vpcId;
    this.baseRanges = props.basePool.map(item => {
      return IPv4CidrRange.fromCidr(item);
    });
  }

  public async init(): Promise<Vpc> {
    await this.vpcDetails();
    await this.subnetDetails();
    return this;
  }

  private async vpcDetails(): Promise<void> {
    // Get VPC details
    const vpcDetails = await throttlingBackOff(() => this.ec2.describeVpcs({ VpcIds: [this.vpcId] }).promise());

    if (!vpcDetails.Vpcs || !vpcDetails.Vpcs[0].CidrBlockAssociationSet) {
      throw new Error(`Unable to retrieve CIDR block details for VPC ${this.vpcId}`);
    }

    // Determine CIDRs allocated by IPAM
    const allocations: IPv4CidrRange[] = [];
    for (const baseRange of this.baseRanges) {
      for (const item of vpcDetails.Vpcs[0].CidrBlockAssociationSet ?? []) {
        if (!item.CidrBlock) {
          throw new Error(`Unable to retrieve CIDR block for VPC ${this.vpcId}`);
        }

        // Compare VPC CIDR to base CIDR range
        const vpcRange = IPv4CidrRange.fromCidr(item.CidrBlock);
        if (vpcRange.inside(baseRange) || vpcRange.isEquals(baseRange)) {
          allocations.push(vpcRange);
        }
      }
    }
    if (allocations.length === 0) {
      throw new Error(`Unable to determine VPC CIDRs allocated by IPAM for VPC ${this.vpcId}`);
    }

    // Get name tag
    const nameTag = vpcDetails.Vpcs[0].Tags?.filter(item => item.Key === 'Name')[0].Value;
    if (!nameTag) {
      throw new Error(`Unable to retrieve name tag for VPC ${this.vpcId}`);
    }

    // Set property values
    this.allocatedCidrs = allocations;
    this.name = nameTag;
  }

  private validateSubnet(subnet: AWS.EC2.Subnet) {
    if (!subnet.SubnetId) {
      throw new Error(`Unable to retrieve subnet ID`);
    }
    if (!subnet.CidrBlock) {
      throw new Error(`Unable to retrieve CIDR block for subnet ${subnet.SubnetId}`);
    }
  }

  private getSubnets(subnetLists: AWS.EC2.SubnetList | undefined): Subnet[] {
    const subnets: Subnet[] = [];

    for (const subnet of subnetLists ?? []) {
      this.validateSubnet(subnet);

      // Determine if subnet is in scope of IPAM
      const subnetCidr = IPv4CidrRange.fromCidr(subnet.CidrBlock!);
      for (const vpcRange of this.allocatedCidrs) {
        if (subnetCidr.inside(vpcRange) || subnetCidr.isEquals(vpcRange)) {
          // Get name tag
          const nameTag = subnet.Tags?.filter(item => item.Key === 'Name')[0].Value;

          // Push object to array
          subnets.push(
            new Subnet({
              allocatedCidr: subnetCidr,
              mapPublicIpOnLaunch: subnet.MapPublicIpOnLaunch ?? false,
              name: nameTag,
              subnetId: subnet.SubnetId!,
              tags: subnet.Tags,
            }),
          );
        }
      }
    }

    return subnets;
  }

  private async subnetDetails(): Promise<void> {
    let nextToken: string | undefined = undefined;
    const subnets: Subnet[] = [];
    do {
      // Get subnet details
      const page = await throttlingBackOff(() =>
        this.ec2
          .describeSubnets({ Filters: [{ Name: 'vpc-id', Values: [this.vpcId] }], NextToken: nextToken })
          .promise(),
      );

      // Iterate through subnets
      subnets.push(...this.getSubnets(page.Subnets));

      nextToken = page.NextToken;
    } while (nextToken);

    // Set property value
    this.subnets = subnets;
  }
}

interface ISubnet {
  /**
   * The CIDR of the subnet allocated by IPAM
   */
  readonly allocatedCidr: IPv4CidrRange;
  /**
   * The ID of the subnet
   */
  readonly subnetId: string;
  /**
   * The MapPublicIpOnLaunch subnet attribute
   */
  readonly mapPublicIpOnLaunch?: boolean;
  /**
   * The friendly name of the subnet
   */
  readonly name?: string;
  /**
   * Tags associated with the subnet
   */
  readonly tags?: AWS.EC2.TagList;
}

interface SubnetProps {
  /**
   * The CIDR of the subnet allocated by IPAM
   */
  readonly allocatedCidr: IPv4CidrRange;
  /**
   * The ID of the subnet
   */
  readonly subnetId: string;
  /**
   * The MapPublicIpOnLaunch subnet attribute
   */
  readonly mapPublicIpOnLaunch?: boolean;
  /**
   * The friendly name of the subnet
   */
  readonly name?: string;
  /**
   * Tags associated with the subnet
   */
  readonly tags?: AWS.EC2.TagList;
}

export class Subnet implements ISubnet {
  public readonly allocatedCidr: IPv4CidrRange;
  public readonly subnetId: string;
  public readonly mapPublicIpOnLaunch: boolean | undefined;
  public readonly name: string | undefined;
  public readonly tags: AWS.EC2.TagList | undefined;

  constructor(props: SubnetProps) {
    this.allocatedCidr = props.allocatedCidr;
    this.mapPublicIpOnLaunch = props.mapPublicIpOnLaunch;
    this.name = props.name;
    this.subnetId = props.subnetId;
    this.tags = props.tags;
  }
}

export async function vpcInit(props: VpcProps): Promise<Vpc> {
  // Set static properties
  const vpc = new Vpc({ basePool: props.basePool, vpcId: props.vpcId });

  // Set VPC and subnet properties
  await vpc.init();
  return vpc;
}
