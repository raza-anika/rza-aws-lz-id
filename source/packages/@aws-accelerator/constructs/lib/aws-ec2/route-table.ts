/**
 *  Copyright 2022 Amazon.com, Inc. or its affiliates. All Rights Reserved.
 *
 *  Licensed under the Apache License, Version 2.0 (the "License"). You may not use this file except in compliance
 *  with the License. A copy of the License is located at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 *  or in the 'license' file accompanying this file. This file is distributed on an 'AS IS' BASIS, WITHOUT WARRANTIES
 *  OR CONDITIONS OF ANY KIND, express or implied. See the License for the specific language governing permissions
 *  and limitations under the License.
 */

import * as cdk from 'aws-cdk-lib';
import { Construct } from 'constructs';

import { PrefixListRoute } from './prefix-list-route';
import { Vpc } from './vpc';

export interface IRouteTable extends cdk.IResource {
  /**
   * The identifier of the route table
   *
   * @attribute
   */
  readonly routeTableId: string;

  /**
   * The VPC associated with the route table
   *
   * @attribute
   */
  readonly vpc: Vpc;
}

export interface RouteTableProps {
  readonly name: string;
  readonly vpc: Vpc;
  readonly tags?: cdk.CfnTag[];
}

export class RouteTable extends cdk.Resource implements IRouteTable {
  public readonly routeTableId: string;

  public readonly vpc: Vpc;

  constructor(scope: Construct, id: string, props: RouteTableProps) {
    super(scope, id);

    this.vpc = props.vpc;

    const resource = new cdk.aws_ec2.CfnRouteTable(this, 'Resource', {
      vpcId: props.vpc.vpcId,
      tags: props.tags,
    });
    cdk.Tags.of(this).add('Name', props.name);

    this.routeTableId = resource.ref;
  }

  public addTransitGatewayRoute(
    id: string,
    transitGatewayId: string,
    transitGatewayAttachment: cdk.CfnResource,
    destination?: string,
    destinationPrefixListId?: string,
    logGroupKmsKey?: cdk.aws_kms.Key,
    logRetentionInDays?: number,
  ): void {
    let route: cdk.aws_ec2.CfnRoute | PrefixListRoute;

    if (destinationPrefixListId) {
      if (!logGroupKmsKey) {
        throw new Error('Attempting to add prefix list route without specifying log group KMS key');
      }
      if (!logRetentionInDays) {
        throw new Error('Attempting to add prefix list route without specifying log group retention period');
      }

      route = new PrefixListRoute(this, id, {
        routeTableId: this.routeTableId,
        destinationPrefixListId,
        logGroupKmsKey,
        logRetentionInDays,
        transitGatewayId,
      });
    } else {
      if (!destination) {
        throw new Error('Attempting to add CIDR route without specifying destination');
      }

      route = new cdk.aws_ec2.CfnRoute(this, id, {
        routeTableId: this.routeTableId,
        destinationCidrBlock: destination,
        transitGatewayId: transitGatewayId,
      });
    }

    route.node.addDependency(transitGatewayAttachment);
  }

  public addNatGatewayRoute(
    id: string,
    natGatewayId: string,
    destination?: string,
    destinationPrefixListId?: string,
    logGroupKmsKey?: cdk.aws_kms.Key,
    logRetentionInDays?: number,
  ): void {
    if (destinationPrefixListId) {
      if (!logGroupKmsKey) {
        throw new Error('Attempting to add prefix list route without specifying log group KMS key');
      }
      if (!logRetentionInDays) {
        throw new Error('Attempting to add prefix list route without specifying log group retention period');
      }

      new PrefixListRoute(this, id, {
        routeTableId: this.routeTableId,
        destinationPrefixListId,
        logGroupKmsKey,
        logRetentionInDays,
        natGatewayId,
      });
    } else {
      if (!destination) {
        throw new Error('Attempting to add CIDR route without specifying destination');
      }

      new cdk.aws_ec2.CfnRoute(this, id, {
        routeTableId: this.routeTableId,
        destinationCidrBlock: destination,
        natGatewayId: natGatewayId,
      });
    }
  }

  public addInternetGatewayRoute(
    id: string,
    destination?: string,
    destinationPrefixListId?: string,
    logGroupKmsKey?: cdk.aws_kms.Key,
    logRetentionInDays?: number,
  ): void {
    if (!this.vpc.internetGateway) {
      throw new Error('Attempting to add Internet Gateway route without an IGW defined.');
    }

    if (!this.vpc.internetGatewayAttachment) {
      throw new Error('Attempting to add Internet Gateway route without an IGW attached.');
    }

    let route: cdk.aws_ec2.CfnRoute | PrefixListRoute;

    if (destinationPrefixListId) {
      if (!logGroupKmsKey) {
        throw new Error('Attempting to add prefix list route without specifying log group KMS key');
      }
      if (!logRetentionInDays) {
        throw new Error('Attempting to add prefix list route without specifying log group retention period');
      }

      route = new PrefixListRoute(this, id, {
        routeTableId: this.routeTableId,
        destinationPrefixListId,
        logGroupKmsKey,
        logRetentionInDays,
        gatewayId: this.vpc.internetGateway.ref,
      });
    } else {
      if (!destination) {
        throw new Error('Attempting to add CIDR route without specifying destination');
      }

      route = new cdk.aws_ec2.CfnRoute(this, id, {
        routeTableId: this.routeTableId,
        destinationCidrBlock: destination,
        gatewayId: this.vpc.internetGateway.ref,
      });
    }

    // Need to add depends on for the attachment, as IGW needs to be part of
    // the network (vpc)
    route.node.addDependency(this.vpc.internetGatewayAttachment);
  }

  public addVirtualPrivateGatewayRoute(
    id: string,
    destination?: string,
    destinationPrefixListId?: string,
    logGroupKmsKey?: cdk.aws_kms.Key,
    logRetentionInDays?: number,
  ): void {
    if (!this.vpc.virtualPrivateGateway) {
      throw new Error('Attempting to add Virtual Private Gateway route without an VGW defined.');
    }
    let route: cdk.aws_ec2.CfnRoute | PrefixListRoute;

    if (destinationPrefixListId) {
      if (!logGroupKmsKey) {
        throw new Error('Attempting to add prefix list route without specifying log group KMS key');
      }
      if (!logRetentionInDays) {
        throw new Error('Attempting to add prefix list route without specifying log group retention period');
      }
      route = new PrefixListRoute(this, id, {
        routeTableId: this.routeTableId,
        destinationPrefixListId,
        logGroupKmsKey,
        logRetentionInDays,
        gatewayId: this.vpc.virtualPrivateGateway.gatewayId,
      });
    } else {
      if (!destination) {
        throw new Error('Attempting to add CIDR route without specifying destination');
      }

      route = new cdk.aws_ec2.CfnRoute(this, id, {
        routeTableId: this.routeTableId,
        destinationCidrBlock: destination,
        gatewayId: this.vpc.virtualPrivateGateway.gatewayId,
      });
    }

    // Need to add depends on for the attachment, as VGW needs to be part of
    // the network (vpc)
    route.node.addDependency(this.vpc.virtualPrivateGatewayAttachment!);
  }

  public addGatewayAssociation(type: string): void {
    if (type === 'internetGateway') {
      const association = new cdk.aws_ec2.CfnGatewayRouteTableAssociation(this, 'GatewayAssociation', {
        routeTableId: this.routeTableId,
        gatewayId: this.vpc.internetGateway!.ref,
      });
      association.node.addDependency(this.vpc.internetGatewayAttachment!);
    }

    if (type === 'virtualPrivateGateway') {
      const association = new cdk.aws_ec2.CfnGatewayRouteTableAssociation(this, 'VirtualPrivateGatewayAssociation', {
        routeTableId: this.routeTableId,
        gatewayId: this.vpc.virtualPrivateGateway!.gatewayId,
      });
      association.node.addDependency(this.vpc.virtualPrivateGatewayAttachment!);
    }
  }
}
