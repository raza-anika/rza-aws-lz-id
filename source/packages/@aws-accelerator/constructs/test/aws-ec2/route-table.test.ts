/**
 *  Copyright 2022 Amazon.com, Inc. or its affiliates. All Rights Reserved.
 *
 *  Licensed under the Apache License, Version 2.0 (the "License"). You may not use this file except in compliance
 *  with the License. A copy of the License is located at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 *  or in the 'license' file accompanying this file. This file is distributed on an 'AS IS' BASIS, WITHOUT WARRANTIES
 *  OR CONDITIONS OF ANY KIND, express or implied. See the License for the specific language governing permissions
 *  and limitations under the License.
 */

import * as cdk from 'aws-cdk-lib';
import { RouteTable } from '../../lib/aws-ec2/route-table';
import { NatGateway, Vpc, Subnet } from '../../lib/aws-ec2/vpc';
import { snapShotTest } from '../snapshot-test';

const testNamePrefix = 'Construct(RouteTable): ';

//Initialize stack for snapshot test and resource configuration test
const stack = new cdk.Stack();
const key = new cdk.aws_kms.Key(stack, 'testKey');
const vpc = new Vpc(stack, 'TestVpc', {
  name: 'Test',
  ipv4CidrBlock: '10.0.0.0/16',
  internetGateway: true,
  enableDnsHostnames: false,
  enableDnsSupport: true,
  instanceTenancy: 'default',
  virtualPrivateGateway: {
    asn: 65000,
  },
});

const rt = new RouteTable(stack, 'RouteTable', {
  name: 'TestRouteTable',
  vpc: vpc,
  tags: [{ key: 'Test-Key', value: 'Test-Value' }],
});

const subnet = new Subnet(stack, 'test-subnet', {
  name: 'test-subnet',
  routeTable: rt,
  vpc,
  availabilityZone: 'a',
  ipv4CidrBlock: '10.0.2.0/24',
});

const ngw = new NatGateway(stack, 'ngw', { name: 'ngw', subnet });

const tgwAttachment = new cdk.aws_ec2.CfnTransitGatewayAttachment(stack, 'tgwAttachment', {
  subnetIds: ['subnet-123'],
  transitGatewayId: 'tgw-12324',
  vpcId: vpc.vpcId,
});

rt.addTransitGatewayRoute('tg-route', 'tgw-1234', tgwAttachment, undefined, 'pl-1234', key, 10);
rt.addTransitGatewayRoute('tg', 'tg-1234', tgwAttachment, '10.0.5.0/24', undefined, key, 10);
rt.addNatGatewayRoute('testNgwRoute', ngw.natGatewayId, '10.0.3.0/24', undefined, key, 10);
rt.addNatGatewayRoute('test2NgwRoute', ngw.natGatewayId, undefined, 'pl-1234', key, 10);
rt.addInternetGatewayRoute('testIgwRoute', '0.0.0.0/0', undefined, key, 10);
rt.addInternetGatewayRoute('testIgwRoute2', undefined, 'pl-1234', key, 10);
rt.addVirtualPrivateGatewayRoute('testVgwRoute', '10.0.30./24', undefined, key, 10);
rt.addVirtualPrivateGatewayRoute('testVgw2Route', undefined, 'pl-1234', key, 10);
rt.addGatewayAssociation('internetGateway');
rt.addGatewayAssociation('virtualPrivateGateway');
/**
 * RouteTable construct test
 */
describe('RouteTable', () => {
  snapShotTest(testNamePrefix, stack);
});
