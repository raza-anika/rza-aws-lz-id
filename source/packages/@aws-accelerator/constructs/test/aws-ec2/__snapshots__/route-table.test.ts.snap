// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`RouteTable Construct(RouteTable):  Snapshot Test 1`] = `
Object {
  "Resources": Object {
    "CustomPrefixListRouteCustomResourceProviderHandler5B28D077": Object {
      "DependsOn": Array [
        "CustomPrefixListRouteCustomResourceProviderRoleD08268B5",
      ],
      "Properties": Object {
        "Code": Object {
          "S3Bucket": Object {
            "Fn::Sub": "cdk-hnb659fds-assets-\${AWS::AccountId}-\${AWS::Region}",
          },
          "S3Key": "REPLACED-GENERATED-NAME.zip",
        },
        "Handler": "__entrypoint__.handler",
        "MemorySize": 128,
        "Role": Object {
          "Fn::GetAtt": Array [
            "CustomPrefixListRouteCustomResourceProviderRoleD08268B5",
            "Arn",
          ],
        },
        "Runtime": "nodejs14.x",
        "Timeout": 900,
      },
      "Type": "AWS::Lambda::Function",
    },
    "CustomPrefixListRouteCustomResourceProviderLogGroup68DB81A5": Object {
      "DeletionPolicy": "Delete",
      "Properties": Object {
        "KmsKeyId": Object {
          "Fn::GetAtt": Array [
            "testKey1CDDDD5E",
            "Arn",
          ],
        },
        "LogGroupName": Object {
          "Fn::Join": Array [
            "",
            Array [
              "/aws/lambda/",
              Object {
                "Ref": "CustomPrefixListRouteCustomResourceProviderHandler5B28D077",
              },
            ],
          ],
        },
        "RetentionInDays": 10,
      },
      "Type": "AWS::Logs::LogGroup",
      "UpdateReplacePolicy": "Delete",
    },
    "CustomPrefixListRouteCustomResourceProviderRoleD08268B5": Object {
      "Properties": Object {
        "AssumeRolePolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": Object {
                "Service": "lambda.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "ManagedPolicyArns": Array [
          Object {
            "Fn::Sub": "arn:\${AWS::Partition}:iam::aws:policy/service-role/AWSLambdaBasicExecutionRole",
          },
        ],
        "Policies": Array [
          Object {
            "PolicyDocument": Object {
              "Statement": Array [
                Object {
                  "Action": Array [
                    "ec2:CreateRoute",
                    "ec2:DeleteRoute",
                  ],
                  "Effect": "Allow",
                  "Resource": "*",
                  "Sid": "AllowModifyRoutes",
                },
              ],
              "Version": "2012-10-17",
            },
            "PolicyName": "Inline",
          },
        ],
      },
      "Type": "AWS::IAM::Role",
    },
    "RouteTable82FB8FA6": Object {
      "Properties": Object {
        "Tags": Array [
          Object {
            "Key": "Name",
            "Value": "TestRouteTable",
          },
          Object {
            "Key": "Test-Key",
            "Value": "Test-Value",
          },
        ],
        "VpcId": Object {
          "Ref": "TestVpcE77CE678",
        },
      },
      "Type": "AWS::EC2::RouteTable",
    },
    "RouteTableGatewayAssociation99E18AA3": Object {
      "DependsOn": Array [
        "TestVpcInternetGatewayAttachment60E451D5",
      ],
      "Properties": Object {
        "GatewayId": Object {
          "Ref": "TestVpcInternetGateway01360C82",
        },
        "RouteTableId": Object {
          "Ref": "RouteTable82FB8FA6",
        },
      },
      "Type": "AWS::EC2::GatewayRouteTableAssociation",
    },
    "RouteTableVirtualPrivateGatewayAssociation6AD07550": Object {
      "DependsOn": Array [
        "TestVpcVirtualPrivateGatewayAttachment5D655F8D",
      ],
      "Properties": Object {
        "GatewayId": Object {
          "Ref": "TestVpcVirtualPrivateGateway56B5C340",
        },
        "RouteTableId": Object {
          "Ref": "RouteTable82FB8FA6",
        },
      },
      "Type": "AWS::EC2::GatewayRouteTableAssociation",
    },
    "RouteTabletest2NgwRoute7F788B8D": Object {
      "DeletionPolicy": "Delete",
      "DependsOn": Array [
        "CustomPrefixListRouteCustomResourceProviderLogGroup68DB81A5",
      ],
      "Properties": Object {
        "ServiceToken": Object {
          "Fn::GetAtt": Array [
            "CustomPrefixListRouteCustomResourceProviderHandler5B28D077",
            "Arn",
          ],
        },
        "routeDefinition": Object {
          "DestinationPrefixListId": "pl-1234",
          "NatGatewayId": Object {
            "Ref": "ngwBD1698D5",
          },
          "RouteTableId": Object {
            "Ref": "RouteTable82FB8FA6",
          },
        },
      },
      "Type": "Custom::PrefixListRoute",
      "UpdateReplacePolicy": "Delete",
    },
    "RouteTabletestIgwRoute23DD8B83D": Object {
      "DeletionPolicy": "Delete",
      "DependsOn": Array [
        "CustomPrefixListRouteCustomResourceProviderLogGroup68DB81A5",
        "TestVpcInternetGatewayAttachment60E451D5",
      ],
      "Properties": Object {
        "ServiceToken": Object {
          "Fn::GetAtt": Array [
            "CustomPrefixListRouteCustomResourceProviderHandler5B28D077",
            "Arn",
          ],
        },
        "routeDefinition": Object {
          "DestinationPrefixListId": "pl-1234",
          "GatewayId": Object {
            "Ref": "TestVpcInternetGateway01360C82",
          },
          "RouteTableId": Object {
            "Ref": "RouteTable82FB8FA6",
          },
        },
      },
      "Type": "Custom::PrefixListRoute",
      "UpdateReplacePolicy": "Delete",
    },
    "RouteTabletestIgwRoute3732DB68": Object {
      "DependsOn": Array [
        "TestVpcInternetGatewayAttachment60E451D5",
      ],
      "Properties": Object {
        "DestinationCidrBlock": "0.0.0.0/0",
        "GatewayId": Object {
          "Ref": "TestVpcInternetGateway01360C82",
        },
        "RouteTableId": Object {
          "Ref": "RouteTable82FB8FA6",
        },
      },
      "Type": "AWS::EC2::Route",
    },
    "RouteTabletestNgwRouteAE321892": Object {
      "Properties": Object {
        "DestinationCidrBlock": "10.0.3.0/24",
        "NatGatewayId": Object {
          "Ref": "ngwBD1698D5",
        },
        "RouteTableId": Object {
          "Ref": "RouteTable82FB8FA6",
        },
      },
      "Type": "AWS::EC2::Route",
    },
    "RouteTabletestVgw2RouteDD19A5B1": Object {
      "DeletionPolicy": "Delete",
      "DependsOn": Array [
        "CustomPrefixListRouteCustomResourceProviderLogGroup68DB81A5",
        "TestVpcVirtualPrivateGatewayAttachment5D655F8D",
      ],
      "Properties": Object {
        "ServiceToken": Object {
          "Fn::GetAtt": Array [
            "CustomPrefixListRouteCustomResourceProviderHandler5B28D077",
            "Arn",
          ],
        },
        "routeDefinition": Object {
          "DestinationPrefixListId": "pl-1234",
          "GatewayId": Object {
            "Ref": "TestVpcVirtualPrivateGateway56B5C340",
          },
          "RouteTableId": Object {
            "Ref": "RouteTable82FB8FA6",
          },
        },
      },
      "Type": "Custom::PrefixListRoute",
      "UpdateReplacePolicy": "Delete",
    },
    "RouteTabletestVgwRoute717D63D3": Object {
      "DependsOn": Array [
        "TestVpcVirtualPrivateGatewayAttachment5D655F8D",
      ],
      "Properties": Object {
        "DestinationCidrBlock": "10.0.30./24",
        "GatewayId": Object {
          "Ref": "TestVpcVirtualPrivateGateway56B5C340",
        },
        "RouteTableId": Object {
          "Ref": "RouteTable82FB8FA6",
        },
      },
      "Type": "AWS::EC2::Route",
    },
    "RouteTabletgCEFADE05": Object {
      "DependsOn": Array [
        "tgwAttachment",
      ],
      "Properties": Object {
        "DestinationCidrBlock": "10.0.5.0/24",
        "RouteTableId": Object {
          "Ref": "RouteTable82FB8FA6",
        },
        "TransitGatewayId": "tg-1234",
      },
      "Type": "AWS::EC2::Route",
    },
    "RouteTabletgroute63CDF5B7": Object {
      "DeletionPolicy": "Delete",
      "DependsOn": Array [
        "CustomPrefixListRouteCustomResourceProviderLogGroup68DB81A5",
        "tgwAttachment",
      ],
      "Properties": Object {
        "ServiceToken": Object {
          "Fn::GetAtt": Array [
            "CustomPrefixListRouteCustomResourceProviderHandler5B28D077",
            "Arn",
          ],
        },
        "routeDefinition": Object {
          "DestinationPrefixListId": "pl-1234",
          "RouteTableId": Object {
            "Ref": "RouteTable82FB8FA6",
          },
          "TransitGatewayId": "tgw-1234",
        },
      },
      "Type": "Custom::PrefixListRoute",
      "UpdateReplacePolicy": "Delete",
    },
    "TestVpcE77CE678": Object {
      "Properties": Object {
        "CidrBlock": "10.0.0.0/16",
        "EnableDnsHostnames": false,
        "EnableDnsSupport": true,
        "InstanceTenancy": "default",
        "Tags": Array [
          Object {
            "Key": "Name",
            "Value": "Test",
          },
        ],
      },
      "Type": "AWS::EC2::VPC",
    },
    "TestVpcInternetGateway01360C82": Object {
      "Properties": Object {
        "Tags": Array [
          Object {
            "Key": "Name",
            "Value": "Test",
          },
        ],
      },
      "Type": "AWS::EC2::InternetGateway",
    },
    "TestVpcInternetGatewayAttachment60E451D5": Object {
      "Properties": Object {
        "InternetGatewayId": Object {
          "Ref": "TestVpcInternetGateway01360C82",
        },
        "VpcId": Object {
          "Ref": "TestVpcE77CE678",
        },
      },
      "Type": "AWS::EC2::VPCGatewayAttachment",
    },
    "TestVpcVirtualPrivateGateway56B5C340": Object {
      "Properties": Object {
        "AmazonSideAsn": 65000,
        "Tags": Array [
          Object {
            "Key": "Name",
            "Value": "Test",
          },
        ],
        "Type": "ipsec.1",
      },
      "Type": "AWS::EC2::VPNGateway",
    },
    "TestVpcVirtualPrivateGatewayAttachment5D655F8D": Object {
      "Properties": Object {
        "VpcId": Object {
          "Ref": "TestVpcE77CE678",
        },
        "VpnGatewayId": Object {
          "Ref": "TestVpcVirtualPrivateGateway56B5C340",
        },
      },
      "Type": "AWS::EC2::VPCGatewayAttachment",
    },
    "ngwBD1698D5": Object {
      "Properties": Object {
        "AllocationId": Object {
          "Fn::GetAtt": Array [
            "ngwEipBD9BC2FE",
            "AllocationId",
          ],
        },
        "SubnetId": Object {
          "Ref": "testsubnetFBD9E7FF",
        },
        "Tags": Array [
          Object {
            "Key": "Name",
            "Value": "ngw",
          },
        ],
      },
      "Type": "AWS::EC2::NatGateway",
    },
    "ngwEipBD9BC2FE": Object {
      "Properties": Object {
        "Domain": "vpc",
        "Tags": Array [
          Object {
            "Key": "Name",
            "Value": "ngw",
          },
        ],
      },
      "Type": "AWS::EC2::EIP",
    },
    "testKey1CDDDD5E": Object {
      "DeletionPolicy": "Retain",
      "Properties": Object {
        "KeyPolicy": Object {
          "Statement": Array [
            Object {
              "Action": "kms:*",
              "Effect": "Allow",
              "Principal": Object {
                "AWS": Object {
                  "Fn::Join": Array [
                    "",
                    Array [
                      "arn:",
                      Object {
                        "Ref": "AWS::Partition",
                      },
                      ":iam::",
                      Object {
                        "Ref": "AWS::AccountId",
                      },
                      ":root",
                    ],
                  ],
                },
              },
              "Resource": "*",
            },
          ],
          "Version": "2012-10-17",
        },
      },
      "Type": "AWS::KMS::Key",
      "UpdateReplacePolicy": "Retain",
    },
    "testsubnetFBD9E7FF": Object {
      "Properties": Object {
        "AvailabilityZone": "a",
        "CidrBlock": "10.0.2.0/24",
        "Tags": Array [
          Object {
            "Key": "Name",
            "Value": "test-subnet",
          },
        ],
        "VpcId": Object {
          "Ref": "TestVpcE77CE678",
        },
      },
      "Type": "AWS::EC2::Subnet",
    },
    "testsubnetRouteTableAssociationA7927A11": Object {
      "Properties": Object {
        "RouteTableId": Object {
          "Ref": "RouteTable82FB8FA6",
        },
        "SubnetId": Object {
          "Ref": "testsubnetFBD9E7FF",
        },
      },
      "Type": "AWS::EC2::SubnetRouteTableAssociation",
    },
    "tgwAttachment": Object {
      "Properties": Object {
        "SubnetIds": Array [
          "subnet-123",
        ],
        "TransitGatewayId": "tgw-12324",
        "VpcId": Object {
          "Ref": "TestVpcE77CE678",
        },
      },
      "Type": "AWS::EC2::TransitGatewayAttachment",
    },
  },
}
`;
